actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,467 @@
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		// this is important: we only add a new entry if there
+		// is not an entry already existing. Therefore, we're only
+		// using the first mapping, satisfying the req for getting the
+		// first matching state for a metastate.
+		if _, ok := metastateToStateMap[metastate]; !ok {
+			metastateToStateMap[metastate] = state
+		}
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes change.Set, attributeName string, oldValue interface{}, newValue interface{}) change.Set {
+	for _, change := range changes {
+		if change.AttributeName == attributeName {
+			change.NewValue = newValue
+			return changes
+		}
+	}
+	newChanges := append(changes, change.Change{
+		AttributeName: attributeName,
+		OldValue:      oldValue,
+		NewValue:      newValue,
+	})
+	return newChanges
+}
+
+func (act ActionStateToMetaState) fuseChanges(c1 change.Set, c2 change.Set) *change.Set {
+	for _, change := range c2 {
+		c1 = act.addOrUpdateChange(c1, change.AttributeName, change.OldValue, change.NewValue)
+	}
+	return &c1
+}
+
+// OnChange executes the action rule. It implements rules.Action.
+func (act ActionStateToMetaState) OnChange(newContext change.Detector, contextChanges change.Set, configuration string, actionChanges *change.Set) (change.Detector, change.Set, error) {
+	if act.Ctx == nil {
+		return nil, nil, errs.New("context is nil")
+	}
+	if act.Db == nil {
+		return nil, nil, errs.New("database is nil")
+	}
+	if act.UserID == nil {
+		return nil, nil, errs.New("userID is nil")
+	}
+	if actionChanges == nil {
+		return nil, nil, errs.New("given actionChanges is nil")
+	}
+	if len(contextChanges) == 0 {
+		// no changes, just return what we have.
+		return newContext, *actionChanges, nil
+	}
+	// if we have multiple changes, this iterates over them and cherrypicks, fusing the results together.
+	var err error
+	var executionChanges change.Set
+	for _, change := range contextChanges {
+		if change.AttributeName == workitem.SystemState {
+			newContext, executionChanges, err = act.onStateChange(newContext, contextChanges, configuration, actionChanges)
+		}
+		if change.AttributeName == workitem.SystemBoardcolumns {
+			newContext, executionChanges, err = act.onBoardColumnsChange(newContext, contextChanges, configuration, actionChanges)
+		}
+		if err != nil {
+			return nil, nil, err
+		}
+		actionChanges = act.fuseChanges(*actionChanges, executionChanges)
+	}
+	// return the result
+	return newContext, *actionChanges, nil
+}
+
+// onBoardColumnsChange is executed when the columns change. It eventually updates the metastate and the state.
+func (act ActionStateToMetaState) onBoardColumnsChange(newContext change.Detector, contextChanges change.Set, configuration string, actionChanges *change.Set) (change.Detector, change.Set, error) {
```

Comment:
Can you please update the comment to include what this function is supposed to return?
This function returns either a workitem object with field values or an empty workitem object based on the value of `wiDirty` flag.

Reply:
See above. Always returns the updated WI. 973fe2d.